pkg/hyper/network.go

```diff
@@ -81,56 +86,368 @@ func toAPIProtocol(protocol Protocol) v1.Protocol {
+			TxQLen: -1,
+		},
+	}
+
+	err := netlink.LinkAdd(br)
+	if err != nil {
+		return nil, fmt.Errorf("could not add %q: %v", brName, err)
+	}
+
+	br, err = bridgeByName(brName)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := netlink.LinkSetUp(br); err != nil {
+		return nil, err
+	}
+
+	return br, nil
 }
 
-func setDownLinksInNs(netns ns.NetNS) error {
-	err := netns.Do(func(_ ns.NetNS) error {
-		links, err := netlink.LinkList()
+func setupRelayBridgeInNs(netns ns.NetNS, cniResult *current.Result) (netlink.Link, error) {
+	var hostVeth netlink.Link
+
+	_, containerLink, err := findContainerLinkInNs(netns, cniResult)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := netns.Do(func(hostNS ns.NetNS) error {
+		var err error
+		var containerVeth netlink.Link
+
+		// setup bridge in ns
+		brName, err := RandomBridgeName()
+		if err != nil {
+			glog.Errorf("Failed to generate bridge name in ns: %v", err)
+			return err
+		}
+
+		br, err := setupBridge(brName)
 		if err != nil {
+			glog.Errorf("Failed to setup bridge in ns: %v", err)
 			return err
 		}
 
-		for _, link := range links {
-			netlink.LinkSetDown(link)
+		// create the veth pair in the container and move host end to host netns
+		vethName, err := RandomVethName()
+		if err != nil {
+			glog.Errorf("Failed to generate veth name in ns: %v", err)
+			return err
+		}
+
+		hostVeth, containerVeth, err = setupVeth(vethName, hostNS)
+		if err != nil {
+			glog.Errorf("Failed to create veth pair in ns: %v", err)
+			return err
+		}
+
+		// connect both new created veth and the old one to the bridge in ns
+		if err := netlink.LinkSetMaster(containerVeth, br); err != nil {
+			glog.Errorf("Failed to connect new created veth to the bridge in ns: %v", err)
+			return err
+		}
+
+		if err := netlink.LinkSetMaster(containerLink, br); err != nil {
+			glog.Errorf("Failed to connect link of container to the bridge in ns: %v", err)
+			return err
 		}
 
 		return nil
-	})
+	}); err != nil {
+		return nil, err
+	}
+
+	return hostVeth, nil
+}
+
+func setupRelayBridgeInHost(hostVeth netlink.Link) (string, error) {
+	// setup bridge in host
+	brName, err := RandomBridgeName()
+	if err != nil {
+		glog.Errorf("Failed to generate bridge name in host: %v", err)
+		return "", err
+	}
+
+	br, err := setupBridge(brName)
 	if err != nil {
+		glog.Errorf("Failed to setup bridge in host: %v", err)
+		return "", err
+	}
+
+	// connect veth to bridge in host
+	if err := netlink.LinkSetMaster(hostVeth, br); err != nil {
+		glog.Errorf("Failed to connect veth to bridge in host: %v", err)
+		return "", err
+	}
+
+	return brName, nil
+}
+
+func teardownRelayBridgeInHost(bridgeName string) error {
+	br, err := bridgeByName(bridgeName)
+	if err != nil {
+		return err
+	}
+
+	if err := netlink.LinkDel(br); err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func buildNetworkInfo(bridgeName string, cniResult *current.Result) *NetworkInfo {
+	ret := &NetworkInfo{}
+
+	ret.BridgeName = bridgeName
```

Comment:
Should also generate the Ifname with same bridge suffix, e.g. for bridge br1a089c6f, generate a tap1a089c6f.